### PR TITLE
[dynamo] fixes dict changed during runtime error 

### LINKED ISSUE #1744: [Bug]: RuntimeError: dictionary changed size during iteration
### 🐛 Describe the bug

Turn off error suppression and then run `test_factory` in `test_aot_cudagraphs.py`. It has a failure that's:

```
  File "/raid/ezyang/pytorch-scratch2/torch/_dynamo/convert_frame.py", line 168, in <listcomp>              
    seen_ids[obj_id] = any([has_tensor(v) for v in obj.__dict__.values()])                         
  File "/raid/ezyang/pytorch-scratch2/torch/_dynamo/convert_frame.py", line 168, in has_tensor                  seen_ids[obj_id] = any([has_tensor(v) for v in obj.__dict__.values()])                                  
  File "/raid/ezyang/pytorch-scratch2/torch/_dynamo/convert_frame.py", line 168, in <listcomp>              
    seen_ids[obj_id] = any([has_tensor(v) for v in obj.__dict__.values()])                                  
  File "/raid/ezyang/pytorch-scratch2/torch/_dynamo/convert_frame.py", line 168, in has_tensor                  seen_ids[obj_id] = any([has_tensor(v) for v in obj.__dict__.values()])                                  
  File "/raid/ezyang/pytorch-scratch2/torch/_dynamo/convert_frame.py", line 168, in <listcomp>              
    seen_ids[obj_id] = any([has_tensor(v) for v in obj.__dict__.values()])         
RuntimeError: dictionary changed size during iteration
```

This shouldn't happen.

### Error logs

na

### Did Dynamo succeed?

- [ ] Does dynamo.optimize("eager") succeed?

### Did AOT succeed?

- [x] Did dynamo.optimize("aot_eager") succeed?

### Did Inductor succeed?

- [x] Does dynamo.optimize("inductor") succeed?

### Minified repro

na

### PR DESCRIPTION
**Pull request by anijain2305**
Fixes https://github.com/pytorch/torchdynamo/issues/1744

cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire @jansel @lezcano @fdrocha

### CODE REVIEW

File: torch/_dynamo/convert_frame.py [hunk 1]
```diff
@@ -156,16 +156,24 @@ def has_tensor(obj):
...
             seen_ids[obj_id] = any([has_tensor(getattr(obj, v)) for v in obj._fields])
             return seen_ids[obj_id]
-        elif not is_allowed(obj) and hasattr(obj, "__dict__") and len(obj.__dict__):
-            seen_ids[obj_id] = any([has_tensor(v) for v in obj.__dict__.values()])
+        elif (
+            not is_allowed(obj)
+            and hasattr(obj, "__dict__")
+            and istype(obj.__dict__, dict)
+        ):
+            seen_ids[obj_id] = has_tensor(obj.__dict__)
```
Comment (ezyang) [inline torch/_dynamo/convert_frame.py:176]: This looks different than what it was before?

File: torch/_dynamo/convert_frame.py [hunk 2]
```diff
@@ -156,16 +156,24 @@ def has_tensor(obj):
             seen_ids[obj_id] = any([has_tensor(v) for v in obj])
             return seen_ids[obj_id]
         elif istype(obj, dict):
-            seen_ids[obj_id] = any([has_tensor(v) for v in obj.values()])
+            # Some packages like py can be updated during runtime. So, make a
```
Comment (ezyang) [inline torch/_dynamo/convert_frame.py:159]: what's py?

File: torch/_dynamo/convert_frame.py [hunk 3]
```diff
@@ -156,16 +156,24 @@ def has_tensor(obj):
...
+            # copy of values to avoid issues like "RuntimeError: dictionary
+            # changed size during iteration"
```
Comment (ezyang) [inline torch/_dynamo/convert_frame.py:161]: This is not much of an explanation at all. Why is the dict getting updated why we traverse over it? Are we triggering user code execution while we iterate over it? Should we prevent that from happening?

File: torch/_dynamo/convert_frame.py (hunk 1 above)
Comment (anijain2305) [reply torch/_dynamo/convert_frame.py:176]: Yeah, the code was duplicated with L159. So, I am redirecting there.

File: torch/_dynamo/convert_frame.py [hunk 4]
```diff
@@ -156,17 +156,25 @@ def has_tensor(obj):
...
-        elif not is_allowed(obj) and hasattr(obj, "__dict__") and len(obj.__dict__):
-            seen_ids[obj_id] = any([has_tensor(v) for v in obj.__dict__.values()])
-            return seen_ids[obj_id]
+        # elif (
+        #     not is_allowed(obj)
+        #     and hasattr(obj, "__dict__")
+        #     and istype(obj.__dict__, dict)
+        # ):
+        #     seen_ids[obj_id] = has_tensor(obj.__dict__)
+        #     return seen_ids[obj_id]
```
Comment (ezyang) [inline torch/_dynamo/convert_frame.py:177]: Don't forget to remove

**Review by ezyang — APPROVED** [review]

**Comment by byronyi** [comment]
This patch seems to trigger infinite recursion with the following repro:

```
pip3 install torch==1.14.0.dev20221110+cpu # 1.14.0.dev20221109+cpu does not reproduce \
 transformers==4.24.0 -f https://download.pytorch.org/whl/nightly/cpu
```

```
import torch
import torch._dynamo

from transformers.tokenization_utils_base import BatchEncoding


@torch._dynamo.optimize('eager')
def tokenization(x):
    encoding = BatchEncoding({'key': x})
    return encoding['key']


x = torch.rand((1, 4))
y = tokenization(x)
```

Stack trace:

```
$ python3 user_dict.py
Traceback (most recent call last):
  File "user_dict.py", line 14, in <module>
    y = tokenization(x)
  File "/home/byronyi/.local/lib/python3.7/site-packages/torch/_dynamo/eval_frame.py", line 169, in _fn
    return fn(*args, **kwargs)
  File "user_dict.py", line 9, in tokenization
    encoding = BatchEncoding({'key': x})
  File "/home/byronyi/.local/lib/python3.7/site-packages/torch/_dynamo/eval_frame.py", line 247, in catch_errors
    return callback(frame, cache_size)
  File "/home/byronyi/.local/lib/python3.7/site-packages/torch/_dynamo/convert_frame.py", line 483, in _convert_frame
    result = inner_convert(frame, cache_size)
  File "/home/byronyi/.local/lib/python3.7/site-packages/torch/_dynamo/convert_frame.py", line 118, in _fn
    return fn(*args, **kwargs)
  File "/home/byronyi/.local/lib/python3.7/site-packages/torch/_dynamo/utils.py", line 87, in time_wrapper
    r = func(*args, **kwargs)
  File "/home/byronyi/.local/lib/python3.7/site-packages/torch/_dynamo/convert_frame.py", line 350, in _convert_frame_assert
    if not has_tensor_in_frame(frame):
  File "/home/byronyi/.local/lib/python3.7/site-packages/torch/_dynamo/eval_frame.py", line 700, in inner_fn
    return fn(*args, **kwargs)
  File "/home/byronyi/.local/lib/python3.7/site-packages/torch/_dynamo/convert_frame.py", line 188, in has_tensor_in_frame
    if has_tensor(value):
  File "/home/byronyi/.local/lib/python3.7/site-packages/torch/_dynamo/convert_frame.py", line 173, in has_tensor
    and not hasattr(obj, "__get__")  # overridden get can mutate the object
  File "/home/byronyi/.local/lib/python3.7/site-packages/transformers/tokenization_utils_base.py", line 248, in __getattr__
    return self.data[item]
  File "/home/byronyi/.local/lib/python3.7/site-packages/transformers/tokenization_utils_base.py", line 248, in __getattr__
    return self.data[item]
  File "/home/byronyi/.local/lib/python3.7/site-packages/transformers/tokenization_utils_base.py", line 248, in __getattr__
    return self.data[item]
  [Previous line repeated 490 more times]
RecursionError: maximum recursion depth exceeded
```

cc @anijain2305 @ezyang

**Comment by anijain2305** [comment]
@byronyi Thanks for reporting the error. I will take a look.

**Comment by anijain2305** [comment]
@byronyi are you sure this PR is the culprit? I see recursion loop even if I revert this PR.

I know the reason why it it loops indefinitely, but its not really related to this. So, I just wanted to double-check with your observation.

**Comment by byronyi** [comment]
> @byronyi are you sure this PR is the culprit? I see recursion loop even if I revert this PR.
> 
> I know the reason why it loops indefinitely, but its not really related to this. So, I just wanted to double-check with your observation.

If not this patch I guess it is another one between 1.14.0.dev20221110+cpu and 1.14.0.dev20221109+cpu as I can't reproduce the issue using the nightly one day before.

**Pull request closed without merging**